core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java

```diff
@@ -167,39 +158,54 @@
     private static final String CONFIG_ROOTS_LIST = "META-INF/quarkus-config-roots.list";
     private static final String[] NO_STRINGS = new String[0];
 
+    static final int DEFAULT_CONVERTER_PRIORITY = 200;
+
     public ConfigurationSetup() {
     }
 
     @BuildStep
-    public void setUpConverters(BuildProducer<ConfigurationCustomConverterBuildItem> configurationTypes) {
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                InetSocketAddress.class,
-                InetSocketAddressConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                CidrAddress.class,
-                CidrAddressConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                InetAddress.class,
-                InetAddressConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                Pattern.class,
-                RegexConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                CipherSuiteSelector.class,
-                CipherSuiteSelectorConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                Protocol.class,
-                ProtocolConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                Path.class,
-                PathConverter.class));
+    public void setUpConverters(BuildProducer<ConfigurationCustomConverterBuildItem> converters) {
+        for (final Converter<?> converter : ServiceLoader.load(Converter.class)) {
+
+            final Class<? extends Converter<?>> converterClass = (Class<? extends Converter<?>>) converter.getClass();
+            final Class<?> returnType = getConverterReturnType(converterClass);
+            final int priority = getConverterPriority(converterClass);
+
+            log.debugf("Register converter %s for %s with priority %s", converterClass, returnType, priority);
+
+            converters.produce(new ConfigurationCustomConverterBuildItem(priority, returnType, converterClass));
+        }
+    }
+
+    /**
+     * Get converter generic return type by reflecting converter class and checking what is returned
+     * by {@link Converter#convert(String)} method.
+     *
+     * @param converterClass the converter class
+     * @return The {@link Converter} output type
+     */
+    static Class<?> getConverterReturnType(final Class<? extends Converter<?>> converterClass) {
```

Comment:
The SmallRye config implementation uses a different approach to determine the converter type:

```java
    private Type getConverterType(Class clazz) {
        if (clazz.equals(Object.class)) {
            return null;
        }

        for (Type type : clazz.getGenericInterfaces()) {
            if (type instanceof ParameterizedType) {
                ParameterizedType pt = (ParameterizedType) type;
                if (pt.getRawType().equals(Converter.class)) {
                    Type[] typeArguments = pt.getActualTypeArguments();
                    if (typeArguments.length != 1) {
                        throw new IllegalStateException("Converter " + clazz + " must be parameterized with a single type");
                    }
                    return typeArguments[0];
                }
            }
        }

        return getConverterType(clazz.getSuperclass());
    }
```

Granted they're using raw types in there (unnecessarily), but it's probably worth ensuring that we're doing the same process.  We could even use this exact method (reflectively at first using an approach similar to that employed by `io.quarkus.runtime.configuration.ConverterFactory#getImplicitConverter`, and then later by making that SmallRye Config method be publicly accessible somehow).

Reply:
Hi @dmlloyd, I used `getConverterType` method as requested and put reflection code in the `ConverterFactory` since it looked to be a good place for a dirty code. After changes here are approved I will prepare a PR to SmallRye repo to make `getConverterType` public static or probably move it to https://github.com/smallrye/smallrye-config/blob/70abffdaadbdcf21ddf5c976157c517e29735f8d/implementation/src/main/java/io/smallrye/config/Converters.java or https://github.com/smallrye/smallrye-config/blob/70abffdaadbdcf21ddf5c976157c517e29735f8d/implementation/src/main/java/io/smallrye/config/ImplicitConverters.java class since these seems to be a better for this kind of functionality.

Reply:
This is already done: https://github.com/smallrye/smallrye-config/pull/111

Reply:
Hah! Nice :)